### ISSUE TITLE
DNS split tunneling and domain-search not working together for the same domain suffix

### ISSUE THREAD

**Issue by mjonasz** [systemd/systemd#24262]
### systemd version the issue has been seen with

systemd-251.3-1

### Used distribution

Archlinux

### Linux kernel version used

5.18.16-zen1-1-zen

### CPU architectures issue was seen on

x86_64

### Component

resolvectl, systemd-resolved

### Expected behaviour you didn't see

DNS split tunneling and `domain-search` DHCP-like option should work correctly even when same domain is used for both. https://en.wikipedia.org/wiki/Search_domain

Something like this should work (note: `tun0` is NOT default interface for DNS here):
```
~> resolvectl domain tun0 wikipedia.org ~wikipedia.org
~> resolvectl domain tun0 
Link 93 (tun0): wikipedia.org ~wikipedia.org
~> nslookup en
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
en.wikipedia.org	canonical name = dyna.wikimedia.org.
Name:	dyna.wikimedia.org
Address: 91.198.174.192
Name:	dyna.wikimedia.org
Address: 2620:0:862:ed1a::1
```

### Unexpected behaviour you saw

With command below `*.wikipedia.org` DNS queries are routed through `tun0` interface correctly, but `domain-search` is ignored.
```
~> resolvectl domain tun0 wikipedia.org ~wikipedia.org
~> resolvectl domain tun0
Link 93 (tun0): ~wikipedia.org
~> nslookup en
Server:		127.0.0.53
Address:	127.0.0.53#53

** server can't find en: NXDOMAIN
```

There is workaround to bypass this issue (not nice enough for production):
```
~> resolvectl domain tun0 wikipedia.org ~ikipedia.org
~> resolvectl domain tun0
Link 93 (tun0): wikipedia.org ~ikipedia.org
~> nslookup en
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
en.wikipedia.org	canonical name = dyna.wikimedia.org.
Name:	dyna.wikimedia.org
Address: 91.198.174.192
Name:	dyna.wikimedia.org
Address: 2620:0:862:ed1a::1
```

### Steps to reproduce the problem

1. Have two network interfaces with DNS servers.
2. One of them should be default route for DNS queries.
3. For second interface set routing (split tunneling) for some domain (e.g. `~wikipedia.org`) and add `domain-search` for it (e.g. `wikipedia.org`).
4. Try `nslookup en` to resolve `en.wikipedia.org` (should expand `en` to `en.wikipedia.org` like here https://en.wikipedia.org/wiki/Search_domain).

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
`resolvectl` output before any commands above (note default DNS route is on `enp0` interface)


Global
           Protocols: -LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
    resolv.conf mode: stub
Fallback DNS Servers: 1.1.1.1#cloudflare-dns.com 9.9.9.9#dns.quad9.net
                      8.8.8.8#dns.google 2606:4700:4700::1111#cloudflare-dns.com
                      2620:fe::9#dns.quad9.net 2001:4860:4860::8888#dns.google

Link 2 (enp0)
    Current Scopes: DNS
         Protocols: +DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 192.168.y.1
       DNS Servers: 192.168.y.1

Link 3 (tun0)
    Current Scopes: DNS
         Protocols: -DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 10.x.0.1
       DNS Servers: 10.x.0.1
```
```

**Comment by mjonasz** [comment]
`nslookup` uses `systemd-resolved` in examples provided. So it doesn't make any difference. Problem is only when same domain is used for DNS split tunneling and `domain-search`.

But for clarity, I also verified `resolvectl query` behavior. It works exactly the same.

(`tun0` is not default interface for DNS)
```
#  DNS tunneling for *wikipedia.org set to tun0 (~wikipedia.org part), domain-search set also to wikipedia.org, `en` not resolving
~> sudo resolvectl domain tun0 wikipedia.org ~wikipedia.org
~> resolvectl query en
en: resolve call failed: 'en' not found

#  DNS tunneling for *ikipedia.org set to tun0 (~ikipedia.org part), domain-search set to wikipedia.org, `en` resolved as `en.wikipedia.org` correctly
~> sudo resolvectl domain tun0 wikipedia.org ~ikipedia.org
~> resolvectl query en
en: 91.198.174.192                             -- link: tun0
    2620:0:862:ed1a::1                         -- link: tun0
    (dyna.wikimedia.org)

-- Information acquired via protocol DNS in 32.1ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
```

**Comment by poettering** [comment]
> `nslookup` uses `systemd-resolved` in examples provided. So it doesn't make any difference.

It does. nslookup goes via DNS proto to talk to local stub. And if you do DNS packets then the search domain logic is done client side, i.e. nslookup implements it, not resolved. And it does so depending on /etc/resolv.conf configuraiton alone.

If you use "resolvectl query" otoh then you allow resolved to do suffix appending for you, and things are interface aware then.

**Comment by poettering** [comment]
And please provide the full status output of resolvectl, that shows how you have configured things.

So I am not sure what "DNS split tunneling" is suppsoed to be?

You are always configure the same domain twice on the interface, once in route-only mode, and once wihtout it. What's the point of that? 

Anyway, I am not grokking what the issue is supposed to be here?

Is this about setting a route-only domain x will disable a previous search domain x being configured? That's expected really, you just declared you only want it for routing, not for searching?

**Comment by mjonasz** [comment]
Thank you very much. 

I misinterpreted some options and documentation (coming from `openvpn` vs `systemd-resolved` configuration). Now it makes sense.

When I set domain without `~` it is used for both routing and searching - exactly how I needed. 

`resolvectl` man pages weren't exactly clear that without `~` prefix domain is set for both routing and searching, but with `~` only for routing:
```
The domain command expects valid DNS domains, possibly prefixed with "~", and configures a per-interface search or route-only domain.
```
